Made Worldshaper's Astrolabe aware of block placement context

TheRealWormbo opened the pull request:
- prevents placed leaves from decaying (fixes #4342 )
- waterloggable blocks become waterlogged when appropriate instead of displacing the water sources
- directional block placement (e.g. for logs or torches) depends on the side the player targets (fixes #4260 )
- only places blocks that can exist in the target location (e.g. no floating buttons or torches that would immediately pop off again), adjusting the player-suggested direction if necessary to properly attach the block
- can stack up multiple items for blocks like sea pickles
- properly triggers sculk sensors from block placement

TheRealWormbo committed:
Made Worldshaper's Astrolabe aware of block placement context

- prevents placed leaves from decaying (fixes #4342)
- waterloggable blocks become waterlogged when appropriate instead of displacing the water sources
- directional block placement (e.g. for logs or torches) depends on the side the player targets
- only places blocks that can exist in the target location (e.g. no floating buttons or torches), adjusting the player-suggested direction if necessary to properly attach the block
- can stack up multiple items for blocks like sea pickles
- properly triggers sculk sensors from block placement

williewillus commented on Xplat/src/main/java/vazkii/botania/common/item/AstrolabeItem.java:
```diff
@@ -110,22 +116,28 @@ public boolean placeAllBlocks(ItemStack stack, Player player) {
+		BlockPos coords = ctx.getClickedPos();
 		player.level.setBlockAndUpdate(coords, state);
 		player.level.levelEvent(LevelEvent.PARTICLES_DESTROY_BLOCK, coords, Block.getId(state));
+		player.level.gameEvent(GameEvent.BLOCK_PLACE, coords, GameEvent.Context.of(player, state));
```
hmmm, reading the code here I'm a bit surprised it wasn't migrated to substituteUse. Basically a couple years back I replaced most of the block-placing code in Botania to do a "swap into hand, call vanilla use, swap back", using `PlayerHelper.substituteUse`. I'm guessing I missed this spot.

Ideally, we should be using that here too, because then vanilla code will handle all the post-place logic, firing game events, etc.

TheRealWormbo committed:
Refactored Astrolabe's block placing and added tests

- refactored to make use of `PlayerHelper.substituteUse`
  - as a result, blocks will no longer be placed if a player couldn't place them, even if they could exist in that spot (no floating candles)
  - this is placement only! (still no mass candle cake production, which would be cake blocks detecting candle use instead of player placing them)
- if stackable blocks (such as sea pickles) are targeted with a matching block selected, placed blocks are centered on that block
- looking for matching items and block providers in the player's inventory only happens once per use instead of up to 1+size^2 times
- fixed `NullPointerException` thrown by `XplatAbstractions.sendToPlayer` implementations in test contexts
- added test cases:
  - checking for proper alignment of placement square (horizontal/vertical)
  - checking for proper block state properties (waterlogging, keeping lit candles lit, etc.)
  - checking that replaceable blocks (and only those) are replaced
  - checking that player placement restrictions apply to all blocks (e.g. not placing floating candles)
  - checking that blocks are properly obtained from one or more item stacks or block providers in the player's inventory
  - checking that Astrolabe can be used from the off-hand
  - checking that the proper amount of mana is consumed for the selected size
  - checking that directionally placeable blocks get the correct orientation

TheRealWormbo committed:
Ensure that Astrolabe doesn't do its work "for free"

If blocks are provided by mana-consuming items, calculations for the number of available blocks may be off if the player is short on mana.
As long as preliminary checks suggest that enough blocks should be available, the Astrolabe will now consume its operation mana first and then only place the blocks that can actually be conjured with the remaining mana.

williewillus commented on Xplat/src/main/java/vazkii/botania/common/item/AstrolabeItem.java:
```diff
@@ -96,149 +101,157 @@ public InteractionResultHolder<ItemStack> use(Level worldIn, Player playerIn, @N
-			if (current >= required) {
-				return true;
-			}
+	public static BlockPlaceContext getBlockPlaceContext(Player player, InteractionHand hand, Block blockToPlace) {
```
annotate return @ Nullable

williewillus commented on Xplat/src/main/java/vazkii/botania/client/render/world/AstrolabePreviewHandler.java:
```diff
@@ -53,10 +54,13 @@ public static void onWorldRenderLast(PoseStack ms, RenderBuffers buffers, Level
-
+		Block blockToPlace = AstrolabeItem.getBlock(stack);
+		int size = AstrolabeItem.getSize(stack);
+		BlockPlaceContext ctx = AstrolabeItem.getBlockPlaceContext(player, player.swingingArm, blockToPlace);
```
not sure how correct `swingingArm` is. Just pass in the hand from the caller?

williewillus commented on Xplat/src/main/java/vazkii/botania/client/render/world/AstrolabePreviewHandler.java:
```diff
@@ -53,10 +54,13 @@ public static void onWorldRenderLast(PoseStack ms, RenderBuffers buffers, Level
+		List<BlockPos> coords = AstrolabeItem.getBlocksToPlace(ctx, size);
+		if (ctx != null && AstrolabeItem.hasBlocks(stack, player, coords.size(), blockToPlace)) {
+			BlockState stateForPlacement = blockToPlace.getStateForPlacement(ctx);
+			BlockState state = stateForPlacement != null ? stateForPlacement : blockToPlace.defaultBlockState();
```
I'm pretty sure a null return from getStateForPlacement means the block says "I can't be here"?

TheRealWormbo replied:
That's correct. It was done this way because it didn't check the resulting block state for each block individually. I changed it now to match the actual placement result as closely as possible.

williewillus commented on Xplat/src/main/java/vazkii/botania/common/item/AstrolabeItem.java:
```diff
@@ -96,149 +101,157 @@ public InteractionResultHolder<ItemStack> use(Level worldIn, Player playerIn, @N
 	}
 
-	public static List<BlockPos> getBlocksToPlace(ItemStack stack, Player player) {
+	public static List<BlockPos> getBlocksToPlace(BlockPlaceContext ctx, int size) {
```
naming nit
```suggestion
	public static List<BlockPos> getPlacePositions(BlockPlaceContext ctx, int size) {
```

williewillus commented on Xplat/src/main/java/vazkii/botania/common/item/AstrolabeItem.java:
```diff
@@ -254,7 +267,7 @@ public void displayRemainderCounter(Player player, ItemStack stack) {
 	private boolean setBlock(ItemStack stack, BlockState state) {
 		if (!state.isAir()) {
-			ItemNBTHelper.setCompound(stack, TAG_BLOCKSTATE, NbtUtils.writeBlockState(state));
+			ItemNBTHelper.setCompound(stack, TAG_BLOCKSTATE, NbtUtils.writeBlockState(state.getBlock().defaultBlockState()));
```
maybe add a comment saying we use the blockstate util methods instead of storing the block id directly for legacy/backward compat, might be confusing to future readers

williewillus commented on Xplat/src/main/java/vazkii/botania/test/item/AstrolabeTest.java:
```diff
@@ -0,0 +1,405 @@
+	private ItemStack getAstrolabeForBlockType(final Player player, Block block) {
+		final ItemStack stack = player.getItemInHand(InteractionHand.MAIN_HAND);
+		ItemNBTHelper.setCompound(stack, AstrolabeItem.TAG_BLOCKSTATE,
+				NbtUtils.writeBlockState(block.defaultBlockState()));
```
expose and use util method from AstrolabeItem instead of assuming the save format in the test?

williewillus reviewed (commented):
very nice tests
just some questions/nits
